pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1053 @@
+};
+
+use crate::{
+	errors::{InvestError, RedeemError},
+	types::{
+		InnerRedeemState, InvestState, InvestTransition, RedeemState, RedeemTransition,
+		TokenSwapReason,
+	},
+	CollectedRedemptionTrancheTokens, Config, Error, Event, ForeignInvestmentInfo,
+	ForeignInvestmentInfoOf, InvestmentState, Pallet, RedemptionState, SwapOf, TokenSwapOrderIds,
+};
+
+mod invest;
+mod redeem;
+
+// Hook execution for (partially) fulfilled token swaps which should be consumed
+// by `TokenSwaps`.
+impl<T: Config> StatusNotificationHook for Pallet<T> {
+	type Error = DispatchError;
+	type Id = T::TokenSwapOrderId;
+	type Status = SwapOf<T>;
+
+	fn notify_status_change(
+		id: T::TokenSwapOrderId,
+		status: SwapOf<T>,
+	) -> Result<(), DispatchError> {
+		let info = ForeignInvestmentInfo::<T>::get(id).ok_or(Error::<T>::InvestmentInfoNotFound)?;
+		let reason = info
+			.last_swap_reason
+			.ok_or(Error::<T>::TokenSwapReasonNotFound)?;
+
+		match reason {
+			TokenSwapReason::Investment => {
+				let pre_state = InvestmentState::<T>::get(&info.owner, info.id).unwrap_or_default();
+				let post_state = pre_state
+					.transition(InvestTransition::FulfillSwapOrder(status))
+					.map_err(|e| {
+						// Inner error holds finer granularity but should never occur
+						log::debug!("ForeignInvestment state transition error: {:?}", e);
+						Error::<T>::from(InvestError::FulfillSwapOrder)
+					})?;
+				Pallet::<T>::apply_invest_state_transition(&info.owner, info.id, post_state)
+			}
+			TokenSwapReason::Redemption => {
+				let pre_state = RedemptionState::<T>::get(&info.owner, info.id).unwrap_or_default();
+				let post_state = pre_state
+					.transition(RedeemTransition::FulfillSwapOrder(status))
+					.map_err(|e| {
+						// Inner error holds finer granularity but should never occur
+						log::debug!("ForeignInvestment state transition error: {:?}", e);
+						Error::<T>::from(RedeemError::FulfillSwapOrder)
+					})?;
+				Pallet::<T>::apply_redeem_state_transition(&info.owner, info.id, post_state)
+			}
+		}
+	}
+}
+
+impl<T: Config> ForeignInvestment<T::AccountId> for Pallet<T> {
+	type Amount = T::Balance;
+	type CollectInvestResult = ExecutedForeignCollectInvest<T::Balance>;
+	type CurrencyId = T::CurrencyId;
+	type Error = DispatchError;
+	type InvestmentId = T::InvestmentId;
+
+	#[transactional]
+	fn increase_foreign_investment(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		amount: T::Balance,
+		foreign_currency: T::CurrencyId,
+		pool_currency: T::CurrencyId,
+	) -> Result<(), DispatchError> {
+		let pre_state = InvestmentState::<T>::get(who, investment_id).unwrap_or_default();
+
+		// NOTE: This adds one db read but let's be safe for the MVP, can hopefully be
+		// removed before deploying to production
+		let unprocessed_invest_amount = T::Investment::investment(who, investment_id)?;
+		ensure!(
+			unprocessed_invest_amount == pre_state.get_investing_amount(),
+			DispatchError::Corruption
+		);
```

Comment:
I am not sure `Corruption` is right here, as we will call this directly from `pallet-liquidity-pools` right?

Assume 
* a user has previously invested but NOT collected
* `T::Investments::investment()` returns a value `!= 0`
* Error should be `NeedCollect` or so

Reply:
Ahh, mmh. If there was in increase in between this is not right.

We have the following rule for normal investments
* a user can ONLY change their order if they can NOT collect

I think we need the same here. If a user can collect, meaning intermediately an epoch was executed, then they need to collect first before they can change their order. This would make the error handling and so on easier.

Reply:
[9d3d041](https://github.com/centrifuge/centrifuge-chain/pull/1473/commits/9d3d04157fea4a0f89c77905436fcfdfc53758b4)